Keep focus on Apply button and show confirmation after applying settings

canacechen9 opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
<!-- Use Closes/Fixes/Resolves #xxx to link this PR to the issue it is responding to. -->
Fixes [Issue #18029](https://github.com/nvaccess/nvda/issues/18029).

### Summary of the issue:
Previously, pressing Alt + A (Apply) would shift focus back to the Categories List, which confused users and made it unclear whether settings were saved. The original post proposed to add a notification for the user to confirm.

### Description of user facing changes:
- Keeps focus on the Apply button after pressing it.
- Shows a non-blocking Window with the message “Settings successfully applied”.
- The message automatically disappears after 3 seconds.

### Description of developer facing changes:
- Uses setPostInitFocus to override the default behavior in postInit, avoiding focus reset to the category list.
- Uses wx.TipWindow with wx.CallLater for lightweight, temporary visual feedback.
- No changes to dialog layout or event handling flow.

### Description of development approach:
- Looked into the onApply() method of the base settings dialog.
- Used setPostInitFocus to maintain focus on the Apply button — leveraging NVDA’s existing design.
- Used wx.TipWindow instead of wx.MessageBox to avoid taking focus or requiring user interaction.
- Added a timer (wx.CallLater) to auto-close the message after 3 seconds for minimal user impact.

### Testing strategy:
-Verified that pressing Alt + A (Apply):
    - Applies settings as expected.
    - Keeps focus on the Apply button.
    - Triggers a tooltip that disappears automatically.

Ensured the fix does not affect other behavior or initial dialog focus.

### Known issues with pull request:
The tooltip does not offer an audio cue unless screen reader focus reaches it (expected for a TipWindow).
No localisation added for the message string ("Settings successfully applied.") — could be added in future.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [ ] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [ ] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [ ] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

Commit:
Fixes #18029: Keep focus on Apply button and show confirmation after applying settings

pre-commit-ci[bot] committed:
Pre-commit auto-fix

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -323,6 +323,12 @@ def onApply(self, evt: wx.CommandEvent):
 		"""
 		self.postInit()
+		# Ensure that after applying, focus stays on the Apply button
+		self.setPostInitFocus = lambda: self.FindWindowById(wx.ID_APPLY).SetFocus()
```
Looking at `postInit`, I think we need to change the base case here. Thoughts?

e.g. replace

```py
# when postInit is called without a setPostInitFocus ie because onApply was called
# then set the focus to the listCtrl. This is a good starting point for a "fresh state"
self.catListCtrl.SetFocus()
```

with


```py
# when postInit is called without a setPostInitFocus ie because onApply was called
# then set the focus to the apply button. This is a good starting point for a "fresh state"
self.FindWindowById(wx.ID_APPLY).SetFocus()
```

you might even be able to remove it all together, so the else condition gets removed entirely

seanbudd replied:
Did you try testing this with removing the `else` block entirely? This would make it so that `postInit` should just retain the current focus. I think setting it to `self.catListCtrl.SetFocus` by default is bad behaviour and should be removed.

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -323,6 +323,12 @@ def onApply(self, evt: wx.CommandEvent):
 		This base method should be called to run the postInit method.
 		"""
 		self.postInit()
+		# Ensure that after applying, focus stays on the Apply button
+		self.setPostInitFocus = lambda: self.FindWindowById(wx.ID_APPLY).SetFocus()
+		# Show a floating message notifying user the settings are applied
+		tip = wx.TipWindow(self, "Settings successfully applied.")
+		# Auto-close the tooltip after 3 seconds
+		wx.CallLater(3000, tip.Close)
```
```suggestion
		# Translators: confirmation message for when settings are applied
		tip = ui.message(_("Settings successfully applied."))
```

seanbudd replied:
You might need to delay this

```suggestion
		# Translators: confirmation message for when settings are applied
		core.callLater(100, ui.message, _("Settings successfully applied."))
```

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -323,6 +323,12 @@ def onApply(self, evt: wx.CommandEvent):
 		This base method should be called to run the postInit method.
 		"""
 		self.postInit()
+		# Ensure that after applying, focus stays on the Apply button
+		self.setPostInitFocus = lambda: self.FindWindowById(wx.ID_APPLY).SetFocus()
+		# Show a floating message notifying user the settings are applied
+		tip = wx.TipWindow(self, "Settings successfully applied.")
+		# Auto-close the tooltip after 3 seconds
+		wx.CallLater(3000, tip.Close)
```
a timeout like this is inaccessible - we need to ensure users can use NVDA at any pace.

canacechen9 replied:
Hey, thanks for the feedback!
I was thinking that with a timeout, the user wouldn't have to actively press another button or take any extra action to confirm. As mentioned in the issue thread, some people feel it's unnecessary to add a separate button just to confirm a confirmation message — which I agree with.

That said, if it’s ultimately better to go with a button and remove the timeout, I’m happy to make that change. I just wanted to explain my reasoning. : )

seanbudd reviewed (changes requested):
Hi - thanks for taking this on @canacechen9!

There's a few major issues with the current implementation, but they can be fixed quite easily.
Please let me know if you need any help

canacechen9 committed:
Apply reviewer suggestion: use ui.message instead of wx.TipWindow

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

canacechen9 committed:
Apply reviewer suggestions: fix self.setPostInitFocus, replace wx.TipWindow with ui.message, use translatable string, remove timer

- Moved self.setPostInitFocus to before self.postInit()
- Replaced wx.TipWindow with ui.message to ensure screen reader users receive the message reliably.
- Marked the message string as translatable using _() and added translation comment for localisation.
- Added a 500ms delay using core.callLater to avoid timing conflicts during dialog focus changes.
- Removed timeout behaviour, which is inaccessible to screen reader users.

canacechen9 committed:
add import ui

SaschaCowley commented:
@canacechen9 do you intend to continue working on this PR? We will close this as abandoned if we don't hear back in 2 weeks.
